pkg/controller/etcdproxy/certs.go

```diff
@@ -7,106 +7,311 @@ import (
+				Namespace:   c.config.ControllerNamespace,
+				Annotations: map[string]string{},
+			},
+			Type: v1.SecretTypeTLS,
+			Data: map[string][]byte{},
+		}
+		err = nil
 	}
-	clientSignerCert, _, err := clientSigner.GetPEMBytes() // goes to etcdproxy
 	if err != nil {
 		return err
 	}
 
-	// Write CA certificate to EtcdProxyController CA ConfigMap.
-	err = c.ensureClientCABundle(etcdstorage, clientSignerCert, clientSigner.Certificates[0].NotAfter)
+	var serverCert *certs.Certificate
+	// Check does the annotation contain valid expiry date, and if not, or it is expired, regenerate the certificate.
+	if expiry, ok := serverSecret.Annotations[ProxyCertificateExpiryAnnotation]; ok {
+		certExpiry, err := time.Parse(time.RFC3339, expiry)
+		if err != nil {
+			return err
+		}
+		if certExpiry.Before(time.Now()) {
+			serverCert, err = c.generateServerBundle(etcdstorage)
+			if err != nil {
+				return err
+			}
 
-	// Generate client certificate for each Secret provided.
-	var errs []error
-	for _, secret := range etcdstorage.Spec.ClientCertSecrets {
-		// Generate client certificate/key pair.
-		clientBundle, err := clientSigner.NewClientCertificate(pkix.Name{CommonName: fmt.Sprintf("client-%s-%s",
-			secret.Namespace, secret.Name)},
-			r.Int63n(100000), currentTime)
+			// Write new Server certificate/key pair to the Secret in the controller namespace.
+			serverCertBytes, serverKeyBytes, err := serverCert.GetPEMBytes()
+			if err != nil {
+				return err
+			}
+
+			serverSecret.Annotations = map[string]string{
+				ProxyCertificateExpiryAnnotation: serverCert.Certificates[0].NotAfter.Format(time.RFC3339),
+				ProxyCertificateSignedBy:         serverCert.Certificates[0].Issuer.CommonName,
+			}
+			serverSecret.Data = map[string][]byte{
+				"tls.crt": serverCertBytes,
+				"tls.key": serverKeyBytes,
+			}
+			err = ensureSecret(c.kubeclientset, serverSecret)
+			if err != nil {
+				return err
+			}
+		}
+	} else if !ok { // If the annotation is not present, generate the certificate.
+		serverCert, err = c.generateServerBundle(etcdstorage)
 		if err != nil {
-			errs = append(errs, err)
-			continue
+			return err
 		}
-		clientCertBytes, clientKeyBytes, err := clientBundle.GetPEMBytes() // goes to apiserver
+
+		// Write new Server certificate/key pair to the Secret in the controller namespace.
+		serverCertBytes, serverKeyBytes, err := serverCert.GetPEMBytes()
 		if err != nil {
-			errs = append(errs, err)
-			continue
+			return err
+		}
+
+		serverSecret.Annotations = map[string]string{
+			ProxyCertificateExpiryAnnotation: serverCert.Certificates[0].NotAfter.Format(time.RFC3339),
+			ProxyCertificateSignedBy:         serverCert.Certificates[0].Issuer.CommonName,
+		}
+		serverSecret.Data = map[string][]byte{
+			"tls.crt": serverCertBytes,
+			"tls.key": serverKeyBytes,
 		}
+		err = ensureSecret(c.kubeclientset, serverSecret)
+		if err != nil {
+			return err
+		}
+	}
+
+	if serverCert == nil {
```

Comment:
any reason we can't unconditionally assign this?

Reply:
:+1: